source/agora/common/DNS.d

```diff
@@ -911,13 +913,90 @@ public struct Domain
+
+        Construct an instance of a `Domain` from a previously validated string
+
+        This method will use `v` as the string stored in the domain,
+        after validating it. Failing validation is a programming error.
+        As a result, this method does not allocate.
+
+        Note that domains always end with a '.', and cannot be empty.
+
+        Params:
+          v = The string to store in the Domain
+
+    ***************************************************************************/
+
+    public static inout(Domain) fromSafeString (inout(char)[] v)
+        @safe pure nothrow @nogc
     {
-        if (v.length && v[$-1] == '.')
-            this.value = v.idup;
-        else
-            this.value = cast(string) (v ~ '.');
+        assert(v.length > 0, "Domain instantiated with an empty value");
+        assert(v[$-1] == '.', "Domain instantiated with non-absolute name");
+
+        size_t lastLabelIdx = 0;
+        foreach (idx, char c; v)
+        {
+            // We can't use `splitter` because it's not `nothrow @nogc`
+            if (c != '.') continue;
+            // This assert is also present in `checkLabel` but the error message
+            // is less precise.
+            const chunk = v[lastLabelIdx .. idx];
+            assert(chunk.length < 64,
+                   "Domain instantiated with a label longer than 64 characters");
+            auto errIdx = Domain.checkLabel(chunk);
+            if (errIdx < chunk.length)
+                assert(0, chunk);
+
+            lastLabelIdx = idx + 1;
+        }
+
+        inout(Domain) result = { value: v };
+        return result;
+    }
+
+    /***************************************************************************
+
+        Construct an instance of a `Domain` from user input
+
+        Unlike `fromSafeString`, this method will always allocate.
+        Validation is done on `v`, and an `Exception` is thrown if it fails.
+        As the string can be user input, the terminating dot can be omitted
+        (e.g. 'google.com' is treated as 'google.com.').
+
+        Params:
+          v = A non-empty string representing a domain
+
+    ***************************************************************************/
+
+    public static inout(Domain) fromString (scope inout(char)[] v) @safe /* pure */
+    {
+        ensure(v.length > 0, "Empty string passing to `Domain.fromString`");
+
+        inout str = (inout x) @trusted {
+            return cast(typeof(x)) ((x[$-1] == '.') ? x.dup : (x ~ '.'));
+        }(v);
+
+        // This is a copy for `fromSafeString`, but uses exceptions
+        size_t lastLabelIdx = 0;
+        foreach (idx, char c; str)
+        {
+            if (c != '.') continue;
+            const chunk = str[lastLabelIdx .. idx];
+            ensure(chunk.length < 64,
+                   "Domain instantiated with a label longer than 64 characters");
+            auto errIdx = Domain.checkLabel(chunk);
+            ensure(errIdx >= chunk.length,
```

Comment:
Why `>=`? L#1149 checks with `==`

Reply:
It shouldn't make a diff, but will change.